### ISSUE TITLE
Book outline tab in edit forms visible, altough books settings do not allow it for content types, allows assignement of this contents to books.

### ISSUE THREAD

**Issue by schoenid** [backdrop/backdrop-issues#5626]
## Description of the bug

Books settings `admin/content/book/settings` are used to allow content types as part of book outlines.
Altough settings are used to not allow book outlines for several content types, the content nodes of these do contain a "Book outline" tab in the edit form, which allows it to add the content to books.
![grafik](https://user-images.githubusercontent.com/15383666/168428207-f7fe3052-0ad5-48a2-ae13-56d7bfc0fb51.png)

If such content type nodes, which are not allowed, are set as book nodes, the structure of the web site may become unpredictable and unmanagable.

## Steps To Reproduce

To reproduce the behavior:
1. Have the books module installed.
2. Go to  `admin/content/book/settings`
3. Select only book 'Book page' on 'Content types allowed in book outlines' and 'Content type for child pages', save configuration
4. Add a 'Post' node or 'Page' node and select the 'Book outline' tab while in the edit form.
5. Note the possibility to <create a new book> or to add this node as Book to any existing book-page.

## Actual behavior

In one case, the display of groups were doubled in a view after selecting an existing book-page for a group.
![grafik](https://user-images.githubusercontent.com/15383666/168427375-2e11e359-13e4-449a-b2a8-a707fcaa2cf7.png)
Both Icons 'Musiker Gruppe' are representing the content of a group. The display is created by a simple content view, by title and picture of the content. Both are representig the same content with the same node ID (`node/10/edit?destination=node/10`).
The group is then displayed also as part of the selected book 'Farben' of another group 'Maler-Gruppe'.
![grafik](https://user-images.githubusercontent.com/15383666/168427833-be8c004c-c116-4955-9b38-e7589153d61c.png)
![grafik](https://user-images.githubusercontent.com/15383666/168427946-29a53ea2-0aab-4bd0-99d4-73d00c6d0007.png)

In another case, the display of that group disapeared.

In both cases, the group could not be unassigned, had to be deleted and rebuilt again !!!
So I had several cases with lost content.

## Expected behavior

Instead of allowing an accidential assignment I would expect, that there is NO 'Book outline' tab:
![grafik](https://user-images.githubusercontent.com/15383666/168428184-7a919444-f4f3-4a87-b2e0-1223bf274d4f.png)

The assignement of content to Books without 'Allowance' should never be possible!

## Additional information

- Backdrop CMS version: 1.21.4
- Organic Groups module installed
- Web server and its version: Apache/2.4.38 (Raspbian)
- Browser(s) and their versions: Opera, Chrome and Firefox (all up to date)

**Comment by herbdool** [comment]
If I understand your description, this is actually not a bug, but a long-standing, controversial feature. See https://github.com/backdrop/backdrop-issues/issues/3638

**Comment by stpaultim** [comment]
@schoenid - I think you are talking about the same thing as issue #3638. You might want to close this issue and post some of your comments there (cut and paste). We might be able to get some momentum behind fixing this issue if we focus on one ticket.

**Comment by schoenid** [comment]
No, I see it as a bug, because you can accidentally destroy content, if you select a book for a content, which has not to be a book.

My opinion is, if the content may not be a book, it has to be prohibited, that it can be selcted as a book, which in the actual state is possible. 

To select a book, where it is not allowed, is an illegal operation, which is made.
A user, which has no experience, will potentially do it, if it's possible. And the result will then be a destruction of content, maybe even the content of others.

The result is non-reversible. You can't deselect the content from the book. You have to delete it and rewrite it.

**Comment by schoenid** [comment]
> I think you are talking about the same thing as issue https://github.com/backdrop/backdrop-issues/issues/3638.

It's clearly not the same.

The "Book outline" tab should be visible for everyone, that has the permission to edit books, if (s)he is on a books node.

Please note:
Actual the "Book outline" tab is visible on **EVERY content**, not only books.
And thats dangerous !!!

**Comment by schoenid** [comment]
The examples in the description on top are groups not books. These groups do contain books, but for the books it's not allowed to add groups to books. 

**_Altough it's possible._** 

So the user responsible for the group "Musiker-Gruppe" was trying to find out, what happens if (s)he selects a book on the groups "Book outline" tab. **Attention:** _The group is not a book!!!_

The result: the group site structure was destroyed. The  group "Musiker-Gruppe" showed up on several places, where it never was intended. The group and the book had to be deleted and with it all the content, which was involved.

- The group and the book had to be recreated - for several hours. 
- The users had to be briefed, not to use the "Book outline" tab on groups - but on books. 
- The situation is dangerous. The risk of a mistake is very high! The situation could reapear every time!

**Comment by stpaultim** [comment]
> It's clearly not the same.

@schoenid - I will need to reread the issue again, because I spent time looking over the other issue last night and I think it's about the same thing. But, in case I am wrong we can keep both tickets open. Here is what I think the other ticket is saying.

1) IF a site selects ONLY book pages should be allowed in book outline
2) And IF a user has permission  to "Administer Book Outlines" 
3) THEN that user has the ability to add other types of content to books - even though they should not, based upon #1

I think that most of us agree that this is a bug. BUT, some people have argued that this was an intentional decision and that it is a feature, because some people might like and use this flexibility. They argue that we should not simply remove this option, because some existing sites might like the idea that a site administrator with permissions is able to OVERRIDE the settings and add content to a book - despite the setting saying that content should not be added.

The proposed solution in that ticket is:

1) To avoid removing this possibility that SOME users MIGHT like, we simply add a setting that allows site admins to turn this off. With the new setting, the "Book Outline" tab would be removed from all content types that are not designated as book content. It seems to me, that this would resolve your concerns. 

You might be arguing that this tab should always be removed and that it NOT be an option to show it to admins. This is also a legitimate perspective. Is that the difference or am I missing something bigger?

**Comment by herbdool** [comment]
Having two issues propose different things to the same thing doesn't usually work IMHO. Better to just hash it out in one place. I propose closing this one in favour of the other

**Comment by stpaultim** [comment]
@schoenid - To be clear, I agree with you that the current situation is broken and needs to be fixed. I tend to agree with @herbdool that we can fix your problem in that issue, whether or not we call it a bug is not important. What is important is that we fix it. @herbdool and I have been doing some work on the book module. I'm willing to work on that issue and welcome your feedback. 

One more note in this issue. I have tried to recreate a situation in which accidentally adding content to book and then trying to remove it causes the loss of data. I'm unable to recreate that specific situation. 

Today, I did the following. 

1) I created a site with organic groups and enabled the book module. 
2) I created group A
3) I created book B (using book pages) and I added that book to group A (I added the pages, I don't think you can add a book to a group, can you?).
4) I then "accidentally" added group A to the outline of Book B. 
5) I then removed group A from Book B and removed Book B from Group A. 

At no point was I unable to remove something from a book or did I encounter any loss of data. 

I believe that you did and wanted to see if it might indicate a bug in Organic Groups, but I am unable to recreate that specific situation. Anyway, if we make it more difficult to accidentally add content to books that is not "allowed" to be in books, that should also help solve your problem.

**Comment by schoenid** [comment]
```
1. IF a site selects ONLY book pages should be allowed in book outline
2. And IF a user has permission to "Administer Book Outlines"
3. THEN that user has the ability to add other types of content to books - even though they should not, based upon (1)
```

Maybe this (2) is the point:

![image](https://user-images.githubusercontent.com/15383666/169698632-f6f6f948-7386-4a8b-ba20-db2dbf2edf1e.png)

So the user - Affie - has no rights, to administer book outlines.
![image](https://user-images.githubusercontent.com/15383666/169698689-c209d1bf-4ace-47d5-9758-f476a55aeacd.png)

But he has the Book outline tab, with the right to administer book outlines, even for the group, which is not a book.
![image](https://user-images.githubusercontent.com/15383666/169698891-ed86a050-9320-4db5-b38a-12dc61e07bfe.png)

=Summary:=

A user, which has no rights to administer book outlines, has the ability, to use the "Book outline" tab, which allows him to add content to a book, which is not allowed.

**Comment by schoenid** [comment]
> The proposed solution in that ticket is:
> 
> 1. To avoid removing this possibility that SOME users MIGHT like, we simply add a setting that allows site admins to turn this off. With the new setting, the "Book Outline" tab would be removed from all content types that are not designated as book content. It seems to me, that this would resolve your concerns.

But if turned on, only users with the permission "Administer book outlines" should have this tab on non-book content.

**Comment by schoenid** [comment]
> 5. I then removed group A from Book B and removed Book B from Group A.

So I've tried to remove the group from the book as user with the role <editor>. Because of this, I hat no possibility.
My point of view was:

1. As editor I was able to add the group to a book outline
2. So as editor I should be able too, to remove the group from the book outline

So (1) is possible, (2) is not.

As Editor I don't have the horizontal "OUTLINE" tab, which is necessary, to remove the group from the books outline.
![image](https://user-images.githubusercontent.com/15383666/169701495-af1d4e2e-0a97-4f6c-8737-1475d919266a.png)

But I have the vertical "Book outline" tab, which allows me, to add the group to the books outline.
![image](https://user-images.githubusercontent.com/15383666/169701625-c59e157f-de33-4c6e-8737-6e01d6501ff9.png)

... This makes in my eyes no sense: either both or none ...

> I believe that you did and wanted to see if it might indicate a bug in Organic Groups

Sorry, I just see it as a bug and have not in mind, to try to force something or someone. I just tried to report my problems with my sites.

> At no point was I unable to remove something from a book or did I encounter any loss of data.

I did not realize, that as "Administrator" I will have the horizontal "OUTLINE" tab.
![image](https://user-images.githubusercontent.com/15383666/169701973-2d7a445e-0eab-43a1-bbd5-5ab30501b905.png)

This gives the possibility, to remove the group from the book.
![image](https://user-images.githubusercontent.com/15383666/169702034-201fbf9a-273d-43d5-9af9-206036087c09.png)

But not for ordinary users ...

> ... if we make it more difficult to accidentally add content to books that is not "allowed" to be in books, that should also help solve your problem.

Rigth.

---

Please think about the following problems and consider a solution:

- For Non-Administrators the horizontal "OUTLINE" tab is not visible, but the vertical "Book outline" tab.
- A user, which has no rights to administer book outlines, has the ability, to use the "Book outline" tab, which allows him to add content to a book, which is not allowed.
- A Non-Administrator is able to add any content to a book outline, but is not able, to remove it from the book outline.

**Comment by schoenid** [comment]
> To be clear, I agree with you that the current situation is broken and needs to be fixed. I tend to agree with @herbdool that we can fix your problem in that issue, whether or not we call it a bug is not important. What is important is that we fix it or at least make it better. 

Ok. I'll add a short comment with a summary of this report to issue https://github.com/backdrop/backdrop-issues/issues/3638.

**Comment by schoenid** [comment]
@stpaultim :
> You might be arguing that this tab should always be removed and that it NOT be an option to show it to admins. This is also a legitimate perspective. Is that the difference or am I missing something bigger?

"to show it to admins" - It's visible for ordinary users.

Sorry, it's just to clarify ...

**Comment by stpaultim** [comment]
I believe I have now been able to recreate a bug that is clearly not intended and is different from #3638. Thanks @schoenid for your patience. 

I can recreate this default Backdrop site without any contrib modules after enabling the book module.

1) Create fresh site and enable the book module.
2) Create a Test User with Editor role
3) Give Editor role permissions to add and edit book pages
4) Test User can now create any type of content, but does not have the Book Outline tab on any of them.
5) Give the Editor Role permission to 'Add content and child pages to books'

![image](https://user-images.githubusercontent.com/3144571/169713893-d7766056-ba4b-4a1a-8181-c2b08e8ef9f7.png)

6) Note

![image](https://user-images.githubusercontent.com/3144571/169713916-566f9b11-c800-4d58-9154-cf147fcb96ba.png)

7) Note that Test User now has the "Book outline" tab on Book nodes, but not for any other content type. 

![image](https://user-images.githubusercontent.com/3144571/169713971-f2e32edb-fe19-49b5-816c-38ab01d6478a.png)

8) Now add "Pages" to the list of content types that can be added to books. 

![image](https://user-images.githubusercontent.com/3144571/169713998-391965fa-81d1-4526-b607-55dea9ac54f0.png)

9) Now note that the "Book outline" tab appears for Test User in all content types.

10) Now remove "Pages" from the list of content types that can be added to books and note that the "Book outline" tab still appears in all Content Types for Test User.

- - - - - - - - - - - - - - 

By default, if someone with the Editor Role has permission to "Add content and child pages to Books" and if Books are the only type of content allowed in Book outlines, then the "Book outline" tab only shows up for Editors on Book Nodes. 

If you add any content types to the list of content that is allowed in Book outlines, then the "Book outline" tab shows up for all content types. And if you remove all content types except Books from the list of content that is allowed in Book outlines, Editors still have the "Book outline" tab on all content types. 

There is no consistent logic here and this must be a bug that goes beyond the situation described in #3638.

**Comment by stpaultim** [comment]
Thanks to a suggestion from @docwilmot, I have a PR that seems to work. @schoenid - can you do some additional testing.

**Comment by herbdool** [comment]
@stpaultim thanks for the PR. Tested and it works well for me. I tested just using core (as you mentioned the bug can be recreated without any contrib modules).

**Comment by schoenid** [comment]
> ... can you do some additional testing.

I've tried the modification on my local server.

- Normal users have no longer a "Book outline" tab.
- The tab is still present for administrators.

It works.
Thanks a lot!

**Comment by BWPanda** [comment]
After reading through this issue, my first thought was that the `in_array()` call probably just needs the third ('strict') parameter set to TRUE (@argiepiano also suggested the same thing). See https://www.php.net/manual/en/function.in-array.php for details.

I'd prefer to see that solution used as it fixes the existing code, rather than trying to bypass it.

**Comment by argiepiano** [comment]
But the `array_filter` already takes care of the problem.  I had suggested using the strict parameter as an alternative for using the `array_filter`. `array_filter` is very commonly used in these situations anyway, to filter out null or 0 values.

The issue was RTBC. Personally I think marking as "needs works" for something that had already been tested, reviewed by at least two people, and RTBC'd by a core committer a waste of volunteer resources.

**Comment by herbdool** [comment]
I agree. The two approaches are quite similar. Maybe one is slightly better than the other but what we've got works well enough.

**Comment by BWPanda** [comment]
As an explanation for my position (in case it's helpful), one of Backdrop's principles is "Simplicity":

> Direct implementations will be chosen over abstraction, and **how things work should be immediately clear and easily documentable.**

To me, just checking if a given value is in an array is simple and easily understandable. It's not obvious to me, however, why filtering the array first is necessary and actually solves this problem.

So, IMO, one of two things needs to happen here:
- Either fix the original code by just adding the third parameter, or
- Document why the filter is needed by adding a comment explaining that line.

Since both require updating the PR, I still think the first option is better. But since @herbdool disagrees then maybe he'll agree to at least the second option. I won't change labels again to get into a 'needs work/rtbc' war, but I do think the PR still needs one of those two things to be done first before it can be committed.

**Comment by stpaultim** [comment]
I will leave this up to the core committers to sort out the next steps. I'm willing to do what I need to to move this issue forward, just let me know. 

I would think that a core committer could just add some documentation while committing? 

If you need me to update the PR, I'd appreciate some help with the text.

**Comment by quicksketch** [comment]
I think the choice between using the `$strict` parameter and `array_filter()` makes little difference. I've seen the `array_filter()` approach used frequently when dealing with checkboxes. It's usage is all over core to strip out the `0` values that end up in the submitted values (and often then in the config files).

With no strong preference one way or the other, I went ahead and merged https://github.com/backdrop/backdrop/pull/4078 as-is into 1.x and 1.22.x. Great sleuthing @stpaultim on the root cause here.

**Comment by schoenid** [comment]
Thanks a lot.
This is a great help for me and the users of my websites.
